conwayste/src/ui/context.rs

```diff
@@ -0,0 +1,374 @@
+                        format!(".on({:?}, ...) was called while .emit call was in progress for {} widget",
+                        what,
+                        stringify!($widget_name)).into()
+                    })?;
+
+                let handler_vec: &mut Vec<crate::ui::context::Handler>;
+                if let Some(vref) = handlers.get_mut(&what) {
+                    handler_vec = vref;
+                } else {
+                    handlers.insert(what, vec![]);
+                    handler_vec = handlers.get_mut(&what).unwrap();
+                }
+                handler_vec.push(hdlr);
+                Ok(())
+            }
+
+            /// Emit an event -- call all handlers for this event's type (as long as they return NotHandled)
+            fn emit(&mut self, event: &crate::ui::context::Event, uictx: &mut crate::ui::context::UIContext) -> Result<(), Box<dyn std::error::Error>> {
+                use crate::ui::context::Handled::*;
+                if self.$handler_data_field.handlers.is_none() {
+                    // save event into forwarded_events for later forwarding
+                    self.$handler_data_field.forwarded_events.push(event.clone());
+                    return Ok(());
+                }
+
+                // take the handlers, so we are not mutably borrowing them more than once during
+                // the call to each handler, below.
+                let mut handlers = self.$handler_data_field.handlers.take().unwrap(); // unwrap OK b/c .is_none() checked above
+
+                // handle regular (non-forwarded) events
+                if let Some(handler_vec) = handlers.get_mut(&event.what) {
+                    // call each handler for this event type, until a Handled is returned
+                    for hdlr in handler_vec {
+                        let handled = hdlr(self, uictx, event)?;
+                        if handled == Handled {
+                            break;
+                        }
+                    }
+                }
+
+                // handle forwarded events
+                loop {
+                    let mut events = vec![];
+                    std::mem::swap(&mut events, &mut self.$handler_data_field.forwarded_events);
+                    if events.len() == 0 {
+                        break;
+                    }
+
+                    for event in events {
+                        if let Some(handler_vec) = handlers.get_mut(&event.what) {
+                            // call each handler for this event type, until a Handled is returned
+                            for hdlr in handler_vec {
+                                let handled = hdlr(self, uictx, &event)?;
+                                if handled == Handled {
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
```

Comment:
Is it possible to get stuck in this loop for ever? The use case I'm thinking -- and this is based off a scan of your phenomenal changes these past few commits -- is if a widget continues to forward events to itself while `Unhandled` is returned due to an implementation bug. If so, what do you think about logging something, for monitoring purposes, to see if we loop some `n` number of times and still have events to process.

Reply:
It's not possible to get stuck in this loop from only returning Unhandled. The inner loop finishes when all the handlers have been called.

Reply:
I wrongly phrased my original question. Is it possible for `events.len() != 0` if a poorly written implementation continues to forward events to itself in one or more of it's handlers?

Reply:
Bump!